src/compiler/checker.ts

```diff
@@ -3697,12 +3707,14 @@ module ts {
                     var maybeCache = maybeStack[depth];
                     // If result is definitely true, copy assumptions to global cache, else copy to next level up
                     var destinationCache = result === Ternary.True || depth === 0 ? relation : maybeStack[depth - 1];
-                    copyMap(/*source*/maybeCache, /*target*/destinationCache);
+                    for (var key in maybeCache) {
+                        destinationCache[key] = maybeCache[key] ? RelationComparisonResult.Succeeded : RelationComparisonResult.Failed;

```

Comment:
Incorrect code here -- `destinationCache` is either `Map<boolean>` or `Map<RelationComparisonResult>`; note that a write of a `RelationComparisonResult` is an unsound write (and incorrect in this case)